src/test/librados/aio_cxx.cc

```diff
@@ -2466,3 +2466,110 @@ TEST(LibRadosAio, MultiReads) {
     ASSERT_EQ(0, memcmp(buf, bl.c_str(), sizeof(buf)));
   }
 }
+
+TEST(LibRadosAio, ReadOperationCancel)
+{
+  Rados cluster;
+  auto pool_prefix = fmt::format("{}_", ::testing::UnitTest::GetInstance()->current_test_info()->name());
+  std::string pool_name = get_temp_pool_name(pool_prefix);
+  ASSERT_EQ("", create_one_pool_pp(pool_name, cluster));
+
+  auto pool_cleanup = make_scope_guard([&] {
+        destroy_one_pool_pp(pool_name, cluster);
+      });
+
+  IoCtx ioctx;
+  cluster.ioctx_create(pool_name.c_str(), ioctx);
+
+  auto c = std::unique_ptr<AioCompletion>{Rados::aio_create_completion()};
+  ObjectReadOperation op;
+  op.assert_exists();
+  ioctx.aio_operate("test_obj", c.get(), &op, nullptr);
+
+  c->cancel();
+  {
+    TestAlarm alarm;
+    ASSERT_EQ(0, c->wait_for_complete());
+  }
+  EXPECT_EQ(-ECANCELED, c->get_return_value());
```

Comment:
I wonder if this is racy.  Going by
```
    /// Request immediate cancellation with error code -ECANCELED
    /// if the operation hasn't already completed.
```
description (which matches my understanding of how existing APIs work), if the op manages to complete before `c->cancel()` is called, wouldn't `c->get_return_value()` return whatever value it completed with instead of the expected `-ECANCELED`?

Reply:
yeah, these tests are inherently racy. i expect the osd round-trip latency to be much larger than the scheduling latency between aio_operate() and cancellation, but that doesn't guarantee we'll never see a failure here

perhaps there's some way to prevent Objecter from actually sending the op? for example by taking the target osd down or using [MOSDBackoff](https://github.com/ceph/ceph/blob/66bf56ec50d0e1688ffc2601a12a365d94466b90/src/osdc/Objecter.cc#L3261-L3280). but i can't think of anything that's reasonable to do within the context of these librados tests

i could drop these test cases, but i tend to think that cancellation tests that pass _most_ of the time are better than no tests at all

Reply:
> i expect the osd round-trip latency to be much larger

writes will have an extra round trip to the replica osds too. doing the reads against an EC pool would also (i think?) add a round trip to make this more reliable

Reply:
interesting idea with pool quotas, but i don't think that helps for reads. i went with the retry loops instead, thanks